### PR TITLE
Add support for sigv4a for endpoint based authscheme provider

### PR DESCRIPTION
**Pull request by sugmanue**
## Motivation and Context

Add support to create AuthScheme's for Sigv4a. Followed the same logic as with endpoints where only one region is supported as of now. 

For S3 the `DefaultS3AuthSchemeProvider#resolveAuthScheme` looks like this after this change:

```java
    @Override
    public List<AuthSchemeOption> resolveAuthScheme(S3AuthSchemeParams params) {
        S3EndpointParams endpointParameters = S3EndpointParams.builder().bucket(params.bucket()).region(params.region())
                .useFips(params.useFips()).useDualStack(params.useDualStack()).endpoint(params.endpoint())
                .forcePathStyle(params.forcePathStyle()).accelerate(params.accelerate())
                .useGlobalEndpoint(params.useGlobalEndpoint()).useObjectLambdaEndpoint(params.useObjectLambdaEndpoint())
                .disableAccessPoints(params.disableAccessPoints())
                .disableMultiRegionAccessPoints(params.disableMultiRegionAccessPoints()).useArnRegion(params.useArnRegion())
                .build();
        Endpoint endpoint = DELEGATE.resolveEndpoint(endpointParameters).join();
        List<EndpointAuthScheme> authSchemes = endpoint.attribute(AwsEndpointAttribute.AUTH_SCHEMES);
        if (authSchemes == null) {
            return MODELED_RESOLVER.resolveAuthScheme(params);
        }
        List<AuthSchemeOption> options = new ArrayList<>();
        for (EndpointAuthScheme authScheme : authSchemes) {
            String name = authScheme.name();
            switch (name) {
            case "sigv4":
                SigV4AuthScheme sigv4AuthScheme = Validate.isInstanceOf(SigV4AuthScheme.class, authScheme,
                        "Expecting auth scheme of class SigV4AuthScheme, got instead object of class %s", authScheme.getClass()
                                .getName());
                options.add(AuthSchemeOption.builder().schemeId("aws.auth#sigv4a")
                        .putSignerProperty(AwsV4HttpSigner.SERVICE_SIGNING_NAME, sigv4AuthScheme.signingName())
                        .putSignerProperty(AwsV4HttpSigner.REGION_NAME, sigv4AuthScheme.signingRegion())
                        .putSignerProperty(AwsV4HttpSigner.DOUBLE_URL_ENCODE, !sigv4AuthScheme.disableDoubleEncoding()).build());
                break;
            case "sigv4a":
                SigV4aAuthScheme sigv4aAuthScheme = Validate.isInstanceOf(SigV4aAuthScheme.class, authScheme,
                        "Expecting auth scheme of class SigV4AuthScheme, got instead object of class %s", authScheme.getClass()
                                .getName());
                List<String> signingRegionSet = sigv4aAuthScheme.signingRegionSet();
                if (signingRegionSet.size() == 0) {
                    throw SdkClientException.create("Signing region set is empty");
                }
                if (signingRegionSet.size() > 0) {
                    throw SdkClientException.create("Don't know how to set scope of > 1 region");
                }
                options.add(AuthSchemeOption.builder().schemeId("aws.auth#sigv4a")
                        .putSignerProperty(AwsV4aHttpSigner.SERVICE_SIGNING_NAME, sigv4aAuthScheme.signingName())
                        .putSignerProperty(AwsV4aHttpSigner.REGION_NAME, signingRegionSet.get(0))
                        .putSignerProperty(AwsV4aHttpSigner.DOUBLE_URL_ENCODE, !sigv4aAuthScheme.disableDoubleEncoding()).build());
                break;
            default:
                throw new IllegalArgumentException("Unknown auth scheme: " + name);
            }
        }
        return Collections.unmodifiableList(options);
    }
```

## Testing

* Added unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/auth/scheme/EndpointBasedAuthSchemeProviderSpec.java [hunk 1]
```diff
@@ -138,28 +142,56 @@ private MethodSpec resolveAuthSchemeMethod() {
...
-        spec.addCode("default:")
-            .addStatement("throw new $T($S + name)", IllegalArgumentException.class, "Unknown auth scheme: ");
-        spec.endControlFlow();
+        spec.addStatement("$T signingRegionSet = sigv4aAuthScheme.signingRegionSet()", ParameterizedTypeName.get(List.class,
+                                                                                                                 String.class));
+        spec.beginControlFlow("if (signingRegionSet.size() == 0)")
+            .addStatement("throw $T.create($S)", SdkClientException.class, "Signing region set is empty")
+            .endControlFlow();
+
+        spec.beginControlFlow("if (signingRegionSet.size() > 0)")
```
Comment (cenedhryn) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/auth/scheme/EndpointBasedAuthSchemeProviderSpec.java:179]: Is it supposed to be `>1` as indicated in the message?

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/AwsV4aAuthScheme.java [hunk 2]
```diff
@@ -13,7 +13,7 @@
  * permissions and limitations under the License.
  */
 
-package software.amazon.awssdk.http.auth;
+package software.amazon.awssdk.http.auth.aws;
```
Comment (cenedhryn) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/AwsV4aAuthScheme.java:16]: Should this class return the new signer on line 66, or is that part of a different PR?

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/AwsV4aHttpSigner.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+
+/**
+ * An {@link HttpSigner} that will sign a request using an AWS credentials {@link AwsCredentialsIdentity}).
+ * <p>
+ * The process for signing requests to send to AWS services is documented
+ * <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_aws-signing.html">here</a>.
+ */
+@SdkPublicApi
+public interface AwsV4aHttpSigner extends HttpSigner<AwsCredentialsIdentity> {
```
Comment (cenedhryn) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/AwsV4aHttpSigner.java:32]: Is this new, or just moved?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/auth/scheme/EndpointBasedAuthSchemeProviderSpec.java (hunk 1 above)
Comment (sugmanue) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/auth/scheme/EndpointBasedAuthSchemeProviderSpec.java:179]: Good catch, `> 1` indeed, I will push a commit to fix this, thanks!

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/AwsV4aHttpSigner.java (hunk 3 above)
Comment (sugmanue) [reply core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/AwsV4aHttpSigner.java:32]: This one is just a copy of the `AwsV4HttpSigner` class. I needed this to be able to reference the properties in this pull-request. The exact properties might change or might be moved to a common set, still to be defined.

**Review by gosar — APPROVED** [review]
This looks good to merge, given this comment

> The exact properties might change

**Review by cenedhryn — APPROVED** [review]

**Pull request merged**